File: library/src/main/java/de/trbnb/databindingcommands/command/RuleCommand.java
package de.trbnb.databindingcommands.command;

import android.content.Context;
import android.support.annotation.NonNull;

import de.trbnb.databindingcommands.functions.Action;
import de.trbnb.databindingcommands.functions.Predicate;

/**
 * A {@link Command} that determines if it is enabled via a {@link Predicate}.
 * This Predicate, or Rule, is set during initialization.
 *
 * The Predicates result will be cached. A refresh can be triggered by calling {@code onEnabledChanged}.
 */
public class RuleCommand extends BaseCommandImpl {

    private final Predicate enabledRule;

    private boolean cachedEnabledRuleResult;

    /**
     * Creates a new RuleCommand.
     *
     * @param action The initial action that will be run when the Command is executed.
     * @param enabledRule The initial rule that determines if this Command is enabled.
     */
    public RuleCommand(Action<Context> action, @NonNull Predicate enabledRule) {
        super(action);

        this.enabledRule = enabledRule;
        evaluateEnabledRuleResult();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public final boolean isEnabled() {
        return cachedEnabledRuleResult;
    }

    /**
     * This method has to be called when the result of the rule might have changed.
     */
    public final void onEnabledChanged(){
        evaluateEnabledRuleResult();
    }

    private void evaluateEnabledRuleResult(){
        cachedEnabledRuleResult = enabledRule.run();

        triggerEnabledChangedListener();
    }
}

File: library/src/main/java/de/trbnb/databindingcommands/command/RuleCommand.kt
package de.trbnb.databindingcommands.command

/**
 * A [Command] that determines if it is enabled via a predicate.
 * This predicate, or "rule", is set during initialization.

 * The predicates result will be cached. A refresh can be triggered by calling `onEnabledChanged`.
 *
 * @param action The initial action that will be run when the Command is executed.
 * @param enabledRule The initial rule that determines if this Command is enabled.
 */
open class RuleCommand<in P, out R>(action: (P) -> R, private val enabledRule: () -> Boolean) : BaseCommandImpl<P, R>(action) {

    override var isEnabled: Boolean = false

    init {
        evaluateEnabledRuleResult()
    }

    /**
     * This method has to be called when the result of the rule might have changed.
     */
    fun onEnabledChanged() {
        evaluateEnabledRuleResult()
    }

    private fun evaluateEnabledRuleResult() {
        isEnabled = enabledRule()

        triggerEnabledChangedListener()
    }
}
